library/src/main/java/com/devbrackets/android/exomedia/core/source/MediaSourceProvider.kt

```diff
@@ -0,0 +1,114 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.devbrackets.android.exomedia.core.source
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.net.Uri
+import android.os.Build
+import android.os.Handler
+
+import com.devbrackets.android.exomedia.BuildConfig
+import com.devbrackets.android.exomedia.ExoMedia
+import com.devbrackets.android.exomedia.core.source.builder.DefaultMediaSourceBuilder
+import com.devbrackets.android.exomedia.core.source.builder.MediaSourceBuilder
+import com.devbrackets.android.exomedia.util.MediaSourceUtil
+import com.google.android.exoplayer2.source.MediaSource
+import com.google.android.exoplayer2.upstream.TransferListener
+
+/**
+ * Provides the functionality to determine which [MediaSource] should be used
+ * to play a particular URL.
+ */
+class MediaSourceProvider {
+
+    @SuppressLint("DefaultLocale")
+    protected var userAgent = String.format(USER_AGENT_FORMAT, BuildConfig.VERSION_NAME, BuildConfig.VERSION_CODE, Build.VERSION.RELEASE, Build.MODEL)
+
+    fun generate(context: Context, handler: Handler, uri: Uri, transferListener: TransferListener?): MediaSource {
+        val sourceTypeBuilder = findByProviders(uri)
+
+        // If a registered builder wasn't found then use the default
+        val builder = sourceTypeBuilder?.builder ?: DefaultMediaSourceBuilder()
+        return builder.build(context, uri, userAgent, handler, transferListener)
+    }
+
+    class SourceTypeBuilder(val builder: MediaSourceBuilder, val uriScheme: String?, val extension: String?, val looseComparisonRegex: String?)
+
+    companion object {
+        protected val USER_AGENT_FORMAT = "ExoMedia %s (%d) / Android %s / %s"
+
+        protected fun findByProviders(uri: Uri): SourceTypeBuilder? {
+            // Uri Scheme (e.g. rtsp)
+            var sourceTypeBuilder = findByScheme(uri)
+            if (sourceTypeBuilder != null) {
+                return sourceTypeBuilder
+            }
+
+            // Extension
+            sourceTypeBuilder = findByExtension(uri)
+            if (sourceTypeBuilder != null) {
+                return sourceTypeBuilder
+            }
+
+            // Regex
+            sourceTypeBuilder = findByLooseComparison(uri)
+            return sourceTypeBuilder
+
+        }
+
+        protected fun findByScheme(uri: Uri): SourceTypeBuilder? {
+            val scheme = uri.scheme
+            if (scheme == null || scheme.isEmpty()) {
+                return null
+            }
+
+            for (builder in ExoMedia.Data.sourceTypeBuilders) {
+                if (builder.uriScheme != null && builder.uriScheme.equals(scheme, ignoreCase = true)) {
```

Comment:
You could do `scheme.equals(builder.uriScheme, ignoreCase = true)` to avoid the null check